docs/blog/posts/await-me-maybe.md

```diff
@@ -0,0 +1,138 @@
+---
+draft: false
+date: 2023-03-15
+categories:
+  - DevLog
+authors:
+  - willmcgugan
+---
+
+# No-async async with Python
+
+A (reasonable) criticism of async is that it tends to proliferate in your code. In order to `await` something, your functions must be `async` all the way up the call-stack.
+
+Before you know it, you have adopted a policy of "async all the things".
+
+<!-- more -->
+
+Textual is very much an async framework, but doesn't *require* the app developer to use the `async` and `await` keywords. Async in Textual apps is entirely optional. This post is about how Textual accomplishes this async-agnosticism.
+
+!!! info
+
+    See this [example](https://textual.textualize.io/guide/widgets/#attributes-down) from the docs for an async-less Textual app.
+
+
+## An apology
+
+But first, an apology! In a previous post I said Textual "doesn't do any IO of its own". This is not accurate. Textual responds to keys and mouse events (**I**nput) and writes content to the terminal (**O**utput).
+
+Although Textual clearly does do IO, it uses `asyncio` mainly for *concurrency*. It allows each widget to update its part of the screen independently from the rest of the app.
+
+## Await me (maybe)
+
+The first no-async async technique is the "Await me maybe" pattern, a term first coined by [Simon Willison](https://simonwillison.net/2020/Sep/2/await-me-maybe/). This is particularly applicable to callbacks (or in Textual terms, message handlers).
+
+The `await_me_maybe` function below can run a callback that is either a plain old function *or* a coroutine (`async def`). It does this by awaiting the result of the callback if it is awaitable, or simply returning the result if it is not.
+
+
+```python
+import asyncio
+import inspect
+
+
+def plain_old_function():
+    return "Plain old function"
+
+async def async_function():
+    return "Async function"
+
+
+async def await_me_maybe(callback):
+    result = callback()
+    if inspect.isawaitable(result):
+        return await result
+    return result
+
+
+async def run_framework():
+    print(
+        await await_me_maybe(plain_old_function)
+    )
+    print(
+        await await_me_maybe(async_function)
+    )
+
+
+if __name__ == "__main__":
+    asyncio.run(run_framework())
+```
+
+## Optionally awaitable
+
+The "await me maybe" pattern is great when an async framework calls the app's code. The app developer can chose to write async code or not. Things get a little more complicated when the an app wants to call the framework's API. If the API has *asynced all the things*, then it would force the app to do the same.
```

Comment:
```suggestion
The "await me maybe" pattern is great when an async framework calls the app's code. The app developer can choose to write async code or not. Things get a little more complicated when the app wants to call the framework's API. If the API has *asynced all the things*, then it would force the app to do the same.
```